src/app/clusters/service-area-server/service-area-server.cpp

```diff
@@ -0,0 +1,1104 @@
+        return ReadProgress(aEncoder);
+
+    case Attributes::FeatureMap::Id:
+        return aEncoder.Encode(mFeature.Raw());
+
+    default:
+        ChipLogProgress(Zcl, "Service Area: Read unsupported attribute %u", aPath.mAttributeId);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void Instance::InvokeCommand(HandlerContext & handlerContext)
+{
+    ChipLogProgress(Zcl, "%s", __func__);
+    switch (handlerContext.mRequestPath.mCommandId)
+    {
+    case Commands::SelectLocations::Id:
+        return CommandHandlerInterface::HandleCommand<Commands::SelectLocations::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & req) { HandleSelectLocationsCmd(ctx, req); });
+
+    case Commands::SkipCurrentLocation::Id:
+        return CommandHandlerInterface::HandleCommand<Commands::SkipCurrentLocation::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & req) { HandleSkipCurrentLocationCmd(ctx); });
+    }
+}
+
+//*************************************************************************
+// attribute readers
+
+CHIP_ERROR Instance::ReadSupportedLocations(AttributeValueEncoder & aEncoder)
+{
+    if (mDelegate->GetNumberOfSupportedLocations() == 0)
+    {
+        return aEncoder.EncodeNull();
+    }
+
+    return aEncoder.EncodeList([this](const auto & encoder) -> CHIP_ERROR {
+        uint8_t locationIndex = 0;
+        LocationStructureWrapper supportedLocation;
+
+        while (mDelegate->GetSupportedLocationByIndex(locationIndex++, supportedLocation))
+        {
+            ReturnErrorOnFailure(encoder.Encode(supportedLocation));
+        }
+        return CHIP_NO_ERROR;
+    });
+}
+
+CHIP_ERROR Instance::ReadSupportedMaps(AttributeValueEncoder & aEncoder)
+{
+    if (mDelegate->GetNumberOfSupportedMaps() == 0)
+    {
+        return aEncoder.EncodeNull();
+    }
+
+    return aEncoder.EncodeList([this](const auto & encoder) -> CHIP_ERROR {
+        uint32_t mapIndex = 0;
+        MapStructureWrapper supportedMap;
+
+        while (mDelegate->GetSupportedMapByIndex(mapIndex++, supportedMap))
+        {
+            ReturnErrorOnFailure(encoder.Encode(supportedMap));
+        }
+        return CHIP_NO_ERROR;
+    });
+}
+
+CHIP_ERROR Instance::ReadSelectedLocations(AttributeValueEncoder & aEncoder)
+{
+    if (mDelegate->GetNumberOfSelectedLocations() == 0)
+    {
+        return aEncoder.EncodeNull();
+    }
+
+    return aEncoder.EncodeList([this](const auto & encoder) -> CHIP_ERROR {
+        uint32_t locationIndex = 0;
+        uint32_t selectedLocation;
+
+        while (mDelegate->GetSelectedLocationByIndex(locationIndex++, selectedLocation))
+        {
+            ReturnErrorOnFailure(encoder.Encode(selectedLocation));
+        }
+        return CHIP_NO_ERROR;
+    });
+}
+
+CHIP_ERROR Instance::ReadProgress(AttributeValueEncoder & aEncoder)
+{
+    if (mDelegate->GetNumberOfProgressElements() == 0)
+    {
+        return aEncoder.EncodeNull();
```

Comment:
Spec says this is nullable, but does not define when it should be null.  Have a hard time evaluating this for correctness.  Please get the spec fixed?

Reply:
Given that the spec puts a contraint that the attribute should containt at least 1 entry, it follows that 0 entries has to be represented by null. This pattern has been used throught this spec and I can't see that we have explicitly stated this for other attributes. Is there an addition to the spec that you would recommend?

I'm not sure Null is useful in these attributes at all. We are replacing 0 with null. This does not add any extra information but complicates implementations since they have to check if something IsNull and has a value before continuing. @plauric Is there a justification for replacing 0 with null?